modules/migrations/gitea_downloader.go

```diff
@@ -0,0 +1,665 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/migrations/base"
+	"code.gitea.io/gitea/modules/structs"
+
+	gitea_sdk "code.gitea.io/sdk/gitea"
+)
+
+var (
+	_ base.Downloader        = &GiteaDownloader{}
+	_ base.DownloaderFactory = &GiteaDownloaderFactory{}
+)
+
+func init() {
+	RegisterDownloaderFactory(&GiteaDownloaderFactory{})
+}
+
+// GiteaDownloaderFactory defines a gitea downloader factory
+type GiteaDownloaderFactory struct {
+}
+
+// New returns a Downloader related to this factory according MigrateOptions
+func (f *GiteaDownloaderFactory) New(ctx context.Context, opts base.MigrateOptions) (base.Downloader, error) {
+	u, err := url.Parse(opts.CloneAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	baseURL := u.Scheme + "://" + u.Host
+	repoNameSpace := strings.TrimPrefix(u.Path, "/")
+	repoNameSpace = strings.TrimSuffix(repoNameSpace, ".git")
+
+	path := strings.Split(repoNameSpace, "/")
+	if len(path) < 2 {
+		return nil, fmt.Errorf("invalid path")
+	}
+
+	//TODO handle gitea installed in subpath ...
+	repoPath := repoNameSpace
```

Comment:
This will need to be done before merging I think

Reply:
status: wip :D